api/s3/s3.go

```diff
@@ -0,0 +1,285 @@
+			log.Errorf("Failed to get dataset information: %v", err)
+			c.AbortWithStatus(http.StatusInternalServerError)
+		}
+		// TODO: Add real creation date
+		buckets = append(buckets, Bucket{Name: datasetInfo.DatasetID,
+			CreationDate: datasetInfo.CreatedAt})
+	}
+
+	c.XML(http.StatusAccepted, ListAllMyBucketsResult{
+		Buckets: buckets,
+		Owner:   Owner{DisplayName: "", ID: ""},
+	})
+}
+
+// ListObjects respondes to an S3 ListObjects request. This request
+// lists the contents of an S3 bucket. We use this to return dataset content.
+// https://docs.aws.amazon.com/AmazonS3/latest/API/API_ListObjects.html
+func ListObjects(c *gin.Context) {
+	log.Debug("S3 ListObjects request")
+
+	dataset := c.Param("dataset")
+
+	allowed := false
+	for _, known := range middleware.GetDatasets(c) {
+		if dataset == known {
+			allowed = true
+
+			break
+		}
+	}
+	if !allowed {
+		c.AbortWithStatus(http.StatusNotFound)
+
+		return
+	}
+
+	files, err := database.GetFiles(dataset)
+	if err != nil {
+		log.Errorf("Failed getting dataset files: %v", err)
+		c.AbortWithStatus(http.StatusInternalServerError)
+
+		return
+	}
+
+	// We return the full upload path, as file key
+	objects := []Object{}
+	for _, file := range files {
+		lastModified, err := time.Parse(time.RFC3339, file.LastModified)
+		if err != nil {
+			log.Errorf("failed to parse last modified time: %v", err)
+			c.AbortWithStatus(http.StatusInternalServerError)
+
+			return
+		}
+		objects = append(objects, Object{
+			Key:          strings.TrimSuffix(file.FilePath, ".c4gh"),
+			Size:         int(file.DecryptedFileSize),
+			LastModified: lastModified.Format(http.TimeFormat),
+		})
+	}
+
+	// Gin doesn't write the xml header when using c.XML, so we add it manually
+	_, err = c.Writer.Write([]byte(xml.Header))
+	if err != nil {
+		log.Errorf("Failed writing XML header: %v", err)
+		c.AbortWithStatus(http.StatusInternalServerError)
+
+		return
+	}
+	c.XML(http.StatusAccepted, ListBucketResult{
+		Name:     dataset,
+		Contents: objects,
+	})
+}
+
+// GetObject respondes to an S3 GetObject request. This request returns S3
+// objects. This is done by first fetching any file that matches the dataset +
+// filename request from the database and then passing the fileID to the
+// SDA Download function.
+// https://docs.aws.amazon.com/AmazonS3/latest/API/API_GetObject.html
+func GetObject(c *gin.Context) {
+	log.Debugf("S3 GetObject request, context: %v", c.Params)
+
+	// Get file info for the given file path (or abort)
+	fileInfo, err := database.GetDatasetFileInfo(c.Param("dataset"), c.Param("filename")+".c4gh")
+	if err != nil {
+		c.AbortWithStatus(http.StatusInternalServerError)
+
+		return
+	}
+
+	if fileInfo == nil {
+		c.AbortWithStatus(http.StatusNotFound)
+
+		return
+	}
+
+	// Set a param so that Download knows to add S3 headers
+	c.Set("S3", true)
+
+	// set the fileID so that download knows what file to download
+	c.Params = append(c.Params, gin.Param{Key: "fileid", Value: fileInfo.FileID})
```

Comment:
If you want to be picky about consistency in the code this line should actually look like this:
```suggestion
	c.Params = append(c.Params, gin.Param{
		Key:   "fileid",
		Value: fileInfo.FileID,
	})
```